BinaryProtocol-8bit.md

```diff
@@ -2,64 +2,102 @@
+
+> [!NOTE]
 >
 > The Harp Binary Protocol uses Little-Endian byte ordering.
 
-## Harp Message specification
+## Harp Message Format
 
-The Harp Message contains a minimal amount of information to execute a well-defined exchange of data. It follows the structure below.
+All Harp messages MUST follow the structure below, specifying the minimal amount of information for a well-defined exchange of data.
 
-| Harp Message |
-| ------------ |
-| MessageType  |
-| Length       |
-| Address      |
-| Port         |
-| PayloadType  |
-| Payload      |
-| Checksum     |
+| Harp Message   |
+| -------------- |
+| MessageType    |
+| Length         |
+| ExtendedLength |
+| Address        |
+| Port           |
+| PayloadType    |
+| Timestamp      |
+| Payload        |
+| Checksum       |
 
 ### MessageType (1 byte)
 
-Specifies the type of the Harp Message.
+Specifies the type and other operation flags of the Harp message. The structure of this byte MUST follow the below specification:
+
+<table>
+<tr>
+    <th align="center">7</th>
+    <th align="center">6</th>
+    <th align="center">5</th>
+    <th align="center">4</th>
+    <th align="center">3</th>
+    <th align="center">2</th>
+    <th align="center">1</th>
+    <th align="center">0</th>
+</tr>
+<tr>
+    <td align="center">0</td>
+    <td align="center">0</td>
+    <td align="center">0</td>
+    <td align="center">Error</td>
+    <td align="center">0</td>
+    <td align="center">0</td>
+    <td align="center" colspan="2">Type</td>
+</tr>
+</table>
+
+#### Type (2 bits)
+
+Specifies the type of the Harp message.
 
 |   Value   |  Description  |
 | :-------  |  ----------- |
-| 1 (`Read`)  |  Read the content of the register with address [`RegisterAddress`]  |
-| 2 (`Write`) |   Write the content to the register with address [`RegisterAddress`]     |
-| 3 (`Event`) |   Send the content of the register with address [`RegisterAddress`]     |
+| 1 (`Read`)  |  Read the contents of the register with the specified [`Address`](#address-1-byte)  |
+| 2 (`Write`) |   Write the contents to the register with the specified [`Address`](#address-1-byte)     |
+| 3 (`Event`) |   Send the contents of the register with the specified [`Address`](#address-1-byte)     |
+
+#### Error flag (1 bit)
+
+When this bit is set, the message represents an error sent from the Device to the Controller.
 
 ### Length (1 byte)
 
-Contains the number of bytes that are still available and need to be read to complete the Harp message (i.e. number of bytes after the field [`Length`]).
+Specifies the number of bytes (`U8`) after this field that are still available and need to be read to complete the Harp message. If one byte is not enough to express the length of the message, the `Length` field MUST be set to 255. In this case the [`ExtendedLength`](#extendedlength-2-bytes) field MUST be included.
+
+### ExtendedLength (2 bytes, OPTIONAL)
```

Comment:
Honestly, now that I see this, is there really anything that prevents us from making this more bytes to allow for more than 16**2? Given that no client has implemented this yet, we may as well consider just pushing this to 4 bytes?